python/bindings.go

```diff
@@ -0,0 +1,397 @@
+    }
+    ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+    defer cancel()
+    host, err := hostdb.Scan(ctx, addr, hostKey)
+    if setError(id, err) {
+        return nil
+    }
+
+    currentHeight, err := siad.ChainHeight()
+    if setError(id, err) {
+        return nil
+    }
+
+    key := ed25519.NewKeyFromSeed(goBytes(key_ptr, 32))
+    funds, err := scanCurrency(totalFunds)
+    if setError(id, err) {
+        return nil
+    }
+    contract, _, err := proto.FormContract(siad, siad, key, host, funds, currentHeight, currentHeight+types.BlockHeight(duration))
+    if setError(id, err) {
+        return nil
+    }
+    buf := make([]byte, C.sizeof_struct_contract_t)
+    copy(buf[0:32], contract.HostKey().Ed25519())
+    copy(buf[32:64], contract.Revision.ParentID[:])
+    copy(buf[64:96], key[:ed25519.SeedSize])
+    return C.CBytes(buf)
+}
+
+//export us_ll_new_session
+func us_ll_new_session(id unsafe.Pointer, client_p unsafe.Pointer, host_str *C.char, contract *C.struct_contract_t) unsafe.Pointer {
+    clientMu.Lock()
+    defer clientMu.Unlock()
+    siad := loadPtr(client_p).(*renterutil.SiadClient)
+    hostKeyPrefix := C.GoString(host_str)
+
+    hostKey, err := siad.LookupHost(hostKeyPrefix)
+    if setError(id, err) {
+        return nil
+    }
+    addr, err := siad.ResolveHostKey(hostKey)
+    if setError(id, err) {
+        return nil
+    }
+    currentHeight, err := siad.ChainHeight()
+    if setError(id, err) {
+        return nil
+    }
+
+    var c renter.Contract
+    copy(c.ID[:], C.GoBytes(unsafe.Pointer(&contract.id), 32))
+    c.HostKey = hostdb.HostKeyFromPublicKey(C.GoBytes(unsafe.Pointer(&contract.hostKey), 32))
+    c.RenterKey = ed25519.NewKeyFromSeed(C.GoBytes(unsafe.Pointer(&contract.renterKey), 32))
+    session, err := proto.NewSession(addr, hostKey, c.ID, c.RenterKey, currentHeight)
+    if setError(id, err) {
+        return nil
+    }
+    return storePtr(session)
+}
+
+//export us_ll_upload
+func us_ll_upload(id unsafe.Pointer, session_p unsafe.Pointer, buf unsafe.Pointer) unsafe.Pointer {
+    session := loadPtr(session_p).(*proto.Session)
+    var sector [renterhost.SectorSize]byte
+    copy(sector[:], goBytes(buf, renterhost.SectorSize))
+    err := session.Write([]renterhost.RPCWriteAction{{
+              Type: renterhost.RPCWriteActionAppend,
+              Data: sector[:],
+    }})
+    if setError(id, err) {
+        return nil
+    }
+    MerkleRoot := merkle.SectorRoot(&sector)
+    return C.CBytes(MerkleRoot[:])
+}
+
+//export us_ll_download
+func us_ll_download(id unsafe.Pointer, session_p unsafe.Pointer, root unsafe.Pointer, buf unsafe.Pointer, offset C.uint, length C.uint) C.ssize_t {
+    session := loadPtr(session_p).(*proto.Session)
+    var sectorMerkleRoot crypto.Hash
+    copy(sectorMerkleRoot[:], goBytes(root, crypto.HashSize))
+    var b bytes.Buffer
+    dst := bufio.NewWriter(&b)
+    err := session.Read(dst, []renterhost.RPCReadRequestSection{{
```

Comment:
I think the `bufio` is actually unnecessary here. You'd want to use `bufio` when `Write` is an expensive operation, but here we're just writing into a `bytes.Buffer`, which should be very fast.

Also, probably best to call `b.Grow(int(length))` on the buffer so that we don't need to re-allocate it.

Reply:
Is there a way to massage `bytes.Buffer` into an `io.Writer` type? That's what the session Read call expects, `bufio` made it stop complaining.

Reply:
Ah, you just need to pass a pointer to the buffer. Alternatively, you could use `bytes.NewBuffer`, which returns a pointer and lets you pre-allocate up front:
```go
b := bytes.NewBuffer(make([]byte, 0, length))
err := session.Read(&b, ...
```